### ISSUE TITLE
Cached.isAdoptable() = false prevents compile time constant access to ContextReference

### ISSUE THREAD

**Issue by JaroslavTulach** [oracle/graal#7131]
**Describe GraalVM and your environment :**

 - GraalVM version 22.3 and 17.0.7, etc.

**Have you verified this issue still happens when using the latest snapshot?**

Yes, the issue persists.

**Describe the issue**

I am about to attach a Maven project that simulates the Enso usage of [TypesLibrary](https://github.com/enso-org/enso/blob/6534d0a925b72a231e21a6824b37d94536d2bd55/engine/runtime/src/main/java/org/enso/interpreter/runtime/library/dispatch/TypesLibrary.java#L16). 

While working on https://github.com/enso-org/enso/pull/7493/files#diff-edc5a078a479a4e8991efef3f4ae7b1fca281a14b32f0431aaf98c848bd55731R192 I realized that sometimes the `EnsoContext.get` doesn't return _compilation constant_. After a [discussion](https://graalvm.slack.com/archives/CNQSB2DHD/p1691076832287559) with Roland and Christian we concluded it is a Truffle DSL processor bug.

**Code snippet or code repository that reproduces the issue**

[trufflelibrarybug.zip](https://github.com/oracle/graal/files/12261225/trufflelibrarybug.zip)

**Steps to reproduce the issue**

```
$ unzip trufflelibrarybug.zip
$ cd trufflelibrarybug/
trufflelibrarybug$ mvn package
trufflelibrarybug$ less target/generated-sources/annotations/cz/xelfi/test/trufflelibrarybug/DefaultLongExportsGen.java
```

The `Cached` inner class overrides `isAdoptable() = false` and that prevents `DefaultLongExports.getType` to effectively use `ContextReference` as `thisLib` doesn't have a `RootNode`.

**Expected behavior**

In `DefaultLongExports.getType` one should be able to check
```java
assert thisLib.getRootNode() != null;
```
and the assert shall succeed.

**Comment by JaroslavTulach** [comment]
What is the status? Has this been already fixed? In 23.1.1? Later?

**Comment by JaroslavTulach** [comment]
We tried to use the `@Bind("$node")` approach, but it seems to fail https://github.com/enso-org/enso/issues/9477 - we are [slowly returning back](https://github.com/enso-org/enso/commit/380f5d2425a353ac1fa44351f31cf9140155fb29) the `@Cached("1") int ignore` trick. Our [GraalVM version is 21.0.2](https://github.com/enso-org/enso/blob/380f5d2425a353ac1fa44351f31cf9140155fb29/build.sbt#L30) and our [Maven libraries are 23.1.2](https://github.com/enso-org/enso/blob/380f5d2425a353ac1fa44351f31cf9140155fb29/build.sbt#L34).
